src/chainstate/burn/sortition.rs

```diff
@@ -0,0 +1,289 @@
+        }
+
+        // pick the next winner
+        let win_idx_opt = BlockSnapshot::sample_burn_distribution(burn_dist, &VRF_seed, sortition_hash);
+        match win_idx_opt {
+            None => {
+                // no winner 
+                Ok(None)
+            },
+            Some(win_idx) => {
+                // winner!
+                Ok(Some(burn_dist[win_idx].candidate.clone()))
+            }
+        }
+    }
+    
+
+    /// Make a block snapshot from is block's data and the previous block.
+    /// This process will:
+    /// * calculate the new consensus hash
+    /// * calculate the total burn so far
+    /// * calculate the new burn quota
+    /// * determine whether or not we can do a sortition, and if so,
+    /// * carry out the sortition to select the next candidate block.
+    ///
+    /// All of this is rolled into the BlockSnapshot struct.
+    /// 
+    /// Call this *after* you store all of the block's transactions to the burn db.
+    pub fn make_snapshot<'a, A, K>(tx: &mut Transaction<'a>, burnchain: &Burnchain, first_block_height: u64, block_height: u64, block_hash: &BurnchainHeaderHash, burn_dist: &Vec<BurnSamplePoint<A, K>>) -> Result<BlockSnapshot, db_error>
+    where
+        A: Address, 
+        K: PublicKey
+    {
+        // txids for operations stored in this block
+        let txids = BurnDB::<A, K>::get_block_txids(tx, block_height, block_hash)?;
+
+        // NOTE: this only counts burns from leader block commits and user burns that match them.
+        // It ignores user burns that don't match any block.
+        let block_burn_total_opt = BurnSamplePoint::get_total_burns(burn_dist);
+        if block_burn_total_opt.is_none() {
+            return Err(db_error::Overflow);
+        }
+        let block_burn_total = block_burn_total_opt.unwrap();
+
+        let last_burn_quota;
+        let last_sortition_hash;
+        let last_burn_total;
+
+        let next_sortition;
+        let next_burn_quota;
+        let winning_block_txid;
+        let winning_block_burn_hash;
+
+        let last_block_snapshot_opt = BurnDB::<A, K>::get_block_snapshot(tx, block_height - 1)?;
+
+        match last_block_snapshot_opt {
+            Some(prev_snapshot) => {
+                last_burn_total = prev_snapshot.total_burn;
+                last_burn_quota = prev_snapshot.burn_quota;
+                last_sortition_hash = prev_snapshot.sortition_hash;
+            },
+            None =>  {
+                last_burn_total = 0;
+                last_burn_quota = 0;
+                last_sortition_hash = SortitionHash::initial();
+            }
+        };
+        
+        let next_sortition_hash = last_sortition_hash.mix_burn_header(block_hash);
+       
+        // did we burn enough?
+        // TODO do we have to be sticklers about the burn quota?  can't we just leave it alone if it's "close enough" to the last burn quota?
+        if block_burn_total >= last_burn_quota {
```

Comment:
What happens in the event that a block doesn't burn more than the quota? Shouldn't its burns be included in the "burn amount" for the next block (as in the variable burn windowing?)

Reply:
Good catch; forgot to add that.